### ISSUE TITLE
[RISC-V] Miscompile at -O2

### ISSUE THREAD

**Issue by patrick-rivos** [llvm/llvm-project#84350]
Testcase:
```c
int printf(const char *, ...);
long a;
short b = 2;
char c(unsigned char d, unsigned char e) { return e == 0 ?: d % e; }
int main() {
  char f;
  a = 1;
  f = c(0 <= 7, b);
  a ^= f;
  printf("%lX\n", a);
}
```

Commands:
```bash
> clang -O2 red.c -o red.out
> qemu-riscv64 red.out
FFFFFFFFFFFFFFFE
> clang red.c -o red.out
> qemu-riscv64 red.out
0
```

`-opt-bisect-limit` points to `Type Promotion on function (main)`

This is a recent regression.
Known good hash: 083d8aa03aca55b88098a91e41e41a8e321a5721
Known bad hash: 8f79cdd8da97c131ae7d8a3210bb69cb6654903d

Found via fuzzer

**Comment by topperc** [comment]
Can you check if https://github.com/llvm/llvm-project/pull/84232 fixes this?

**Comment by patrick-rivos** [comment]
I checked out the head of the PR branch at https://github.com/llvm/llvm-project/pull/84232 and this issue does not exist there.

```
> clang -O3 red.c -o red.out
> qemu-riscv64 red.out
0
```

Thanks!

**Comment by topperc** [comment]
Thanks for confirming. I'm trying to mitigate some of the fallout from that fix before committing it.

**Comment by dtcxzyw** [comment]
Should be fixed by #84232.
